### PR TITLE
Naive clock jump implementation - allows for clock reuse and simplified Player setup

### PR DESCRIPTION
**Pull request by emersonknapp**
Clock can be created in the constructor so it is never null, and have all its state simply maintained while looping or repeated calls to `play`.

This is an intentionally simple implementation that accomplishes this single goal, and is definitely not safe for use _during_ playback. There is significantly more complexity involved in handling jumping during playback - but that is out of the scope of this change.

### CODE REVIEW

**Comment by emersonknapp** [comment]
@MichaelOrlov FYI

**Comment by emersonknapp** [comment]
Gist: https://gist.githubusercontent.com/emersonknapp/c71f7d18aab7ff09541be9b2d0f22c1b/raw/8401041718d2e42dd2704c81592b62925181b0e2/ros2.repos
BUILD args: --packages-up-to rosbag2_cpp rosbag2_transport rosbag2_tests rosbag2
TEST args: --packages-select rosbag2_cpp rosbag2_transport rosbag2_tests rosbag2
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/8312
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14471/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9238/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12139/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14571/)

**Review by MichaelOrlov — CHANGES_REQUESTED** [review]

File: rosbag2_transport/src/rosbag2_transport/player.cpp [hunk 1]
```diff
@@ -183,27 +197,17 @@ void Player::play()
   try {
     do {
       reader_->open(storage_options_, {"", rmw_get_serialization_format()});
```
[the remark this answers is not in the document]
Comment (emersonknapp) [reply rosbag2_transport/src/rosbag2_transport/player.cpp:199]: `reset` closes the Reader - you have to call `open` again every time after it is `reset`

File: rosbag2_transport/src/rosbag2_transport/player.cpp [hunk 2]
```diff
@@ -183,27 +197,17 @@ void Player::play()
...
+        reader_->get_metadata().starting_time.time_since_epoch()).count();
 
       storage_loading_future_ = std::async(
         std::launch::async,
         [this]() {load_storage_content();});
 
       wait_for_filled_queue();
 
+      clock_->jump(starting_time);
+      clock_->resume();
```
[the remark this answers is not in the document]
Comment (emersonknapp) [reply rosbag2_transport/src/rosbag2_transport/player.cpp:210]: I don't think this makes sense to me - `play` implies a `resume` to me. Messages should always start when you call `play`, right? The scenario you are describing, to me, sounds like

* Create the Player, but don't call `play` - it's "paused"
* call `play` when you want it to start

File: rosbag2_transport/src/rosbag2_transport/player.cpp [hunk 3]
```diff
@@ -96,6 +96,20 @@ Player::Player(
   storage_options_(storage_options),
   play_options_(play_options)
 {
+  {
+    reader_->open(storage_options_, {"", rmw_get_serialization_format()});
+    double rate = play_options_.rate > 0.0 ? play_options_.rate : 1.0;
```
[the remark this answers is not in the document]
Comment (emersonknapp) [reply rosbag2_transport/src/rosbag2_transport/player.cpp:101]: I disagree. `PlayOptions` is passed all the way from the CLI layer, I don't see any reason to force that layer to make multiple calls to configure, when you could configure with a single options struct. Additionally, we want the player to be able to be initialized as a Component Node with Parameters, meaning there is _no way_ to make an additional call in construction, for this configuration.

[the remark this answers is not in the document]
Comment (emersonknapp) [reply rosbag2_transport/src/rosbag2_transport/player.cpp:101]: I will also not want to deal with this in this PR - as this line simple refactoring of _existing_ behavior.

[the remark this answers is not in the document]
Comment (emersonknapp) [reply rosbag2_transport/src/rosbag2_transport/player.cpp:101]: > start clock with rate equal to 1 on construction since we are not going to run clock right after construction anyway. And then adjust rate only via setter method. 

This part is fine - but outside the scope of this change. PRs welcome

File: rosbag2_transport/src/rosbag2_transport/player.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (MichaelOrlov) [reply rosbag2_transport/src/rosbag2_transport/player.cpp:210]: WRT. 
> Messages should always start when you call play, right? 

Nop. I think it could be a situation when someone want to have deferred play. Calling `play(..)`  will prepare everything to play and resume will trigger the real playback immediately. Inside `play(..)` method a lot of initialization routines and even spawning new tread and wait for storage to load content in the queue. \
At least it will be consistent according to the internal state of the player.

There are some situations when it's necessary to minimize timeout when you start playing data after some event happened. \
Consider the case if you have a huge bag file with sequential reader and want to start playing at the timestamp close to the end. It could take some substantial time when reader will adjust his position and preload data in buffers.

File: rosbag2_transport/src/rosbag2_transport/player.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (MichaelOrlov) [reply rosbag2_transport/src/rosbag2_transport/player.cpp:199]: This `reader->reset()` misleading me one more time. I think it's just because the name for this method is not really describe what it's doing. 
even in doxygen description for this method stated that it's really `close` rather than `reset`
```cpp
/**
   * Closing the reader instance.
   */
  void reset();
```
@emersonknapp I think it would be good to rename this method in future. Do you agree? \
Of course in separate PR.

**Review by Karsten1987 — COMMENTED** [review]
I am generally okay with it, just a few comments inline.
Before really approving it, I'd like to @emersonknapp's opinion of tests for this. Are we having tests covering all the new things in this PR?

File: rosbag2_transport/src/rosbag2_transport/player.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Karsten1987) [reply rosbag2_transport/src/rosbag2_transport/player.cpp:199]: I agree, renaming it to `close` or something sounds better to me.

File: rosbag2_transport/src/rosbag2_transport/player.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (MichaelOrlov) [reply rosbag2_transport/src/rosbag2_transport/player.cpp:210]: Ok. Let's keep it as is for now. Otherwise we need to init clock with resume state on constructor.

File: rosbag2_cpp/include/rosbag2_cpp/clocks/time_controller_clock.hpp [hunk 4]
```diff
@@ -102,6 +102,16 @@ class TimeControllerClock : public PlayerClock
   ROSBAG2_CPP_PUBLIC
   bool is_paused() const override;
 
+  /**
+   * Change the current ROS time to an arbitrary time.
+   * This will wake any waiting `sleep_until`.
+   * Note: this initial implementation does not have any jump-callback handling,
+   * so it is not safe to use during playback, only for reconfiguration for clock reuse.
```
[the remark this answers is not in the document]
Comment (emersonknapp) [reply rosbag2_cpp/include/rosbag2_cpp/clocks/time_controller_clock.hpp:109]: > what does "reconfiguration for clock reuse" mean?

I mean - "to reconfigure the clock so that you can use the same clock instance again, with all current settings, at a different starting time". Is that wording any more clear?

>  should we then actively prevent this from happening?

No - because the clock has no concept of "playback". This is a "usage hint" to the `Player`. We could restrict its usage to "only while paused" - but that doesn't really help anything, the `Player`'s current queueing mechanism will still be completely broken if `jump` is used while the queue is active.

File: rosbag2_cpp/include/rosbag2_cpp/clocks/player_clock.hpp [hunk 5]
```diff
@@ -90,6 +90,12 @@ class PlayerClock
    */
   ROSBAG2_CPP_PUBLIC
   virtual bool is_paused() const = 0;
+
+  /**
+   * Change the current ROS time to an arbitrary time.
+   */
+  ROSBAG2_CPP_PUBLIC
+  virtual void jump(rcutils_time_point_value_t) = 0;
```
[the remark this answers is not in the document]
Comment (emersonknapp) [reply rosbag2_cpp/include/rosbag2_cpp/clocks/player_clock.hpp:98]: I agree with this

[the remark this answers is not in the document]
Comment (emersonknapp) [reply rosbag2_cpp/include/rosbag2_cpp/clocks/player_clock.hpp:98]: Oh - actually - semantically this would be a `rclcpp::Time` rather than a `rclcpp::Duration` - same deal though.

File: rosbag2_transport/src/rosbag2_transport/player.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (emersonknapp) [reply rosbag2_transport/src/rosbag2_transport/player.cpp:199]: https://github.com/ros2/rosbag2/pull/760

File: rosbag2_transport/src/rosbag2_transport/player.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (MichaelOrlov) [reply rosbag2_transport/src/rosbag2_transport/player.cpp:210]: I actually recall a real use-case wen we will need playback starting in pause. \
This is "deterministic replay". It's supposed to be used in junction with `play_next()` message.

The idea is that for replay messages deterministically we need to control at what time to play next message. And this control is driven by the receiver side. We need to play next message only when we will get confirmation that all recipients have received previous on.  \
The flow would be: 
1. Create player
2. Put player in pause
3. Call Player::play(..) - player shall not call resume - otherwise it will start playing and we loose determinism in replay.
4. Call play_next() by event i.e. callback from services interface

I would suggest to init clock with `pause` equal to `false` in constructor and remove `resume()` in `player::play(..)`. In this case default behavior will be the same as in current implementation. The normal flow will be:
1. Create player
3. Call Player::play(..) - player shall not call resume. Playback will start automatically because settled `pause == false` in constructor

At the same time we will have opportunity for deterministic replay scenario with `play_next()` calling `Player::pause()` right before `Player::play()`.

[the remark this answers is not in the document]
Comment (emersonknapp) [reply rosbag2_transport/src/rosbag2_transport/player.cpp:210]: Ok - that makes sense to me, I will change this

File: rosbag2_cpp/include/rosbag2_cpp/clocks/player_clock.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (emersonknapp) [reply rosbag2_cpp/include/rosbag2_cpp/clocks/player_clock.hpp:98]: I've decided to just do this in this PR :+1:

File: rosbag2_transport/src/rosbag2_transport/player.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (emersonknapp) [reply rosbag2_transport/src/rosbag2_transport/player.cpp]: The latest revision should reflect this desired behavior.

File: rosbag2_transport/src/rosbag2_transport/player.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (emersonknapp) [reply rosbag2_transport/src/rosbag2_transport/player.cpp]: Updated to remove rate from the clock constructor and add the check, for now, in the clock implemetnation.

**Comment by emersonknapp** [comment]
@Karsten1987 @MichaelOrlov let me know how you feel about the latest revision - it's maybe a little overloaded (doing a couple different things), but since this is a "reconfigure some things for ease of use" PR I am thinking that's probably OK

File: rosbag2_cpp/include/rosbag2_cpp/clocks/player_clock.hpp [hunk 6]
```diff
@@ -60,13 +61,21 @@ class PlayerClock
   virtual bool sleep_until(rcutils_time_point_value_t until) = 0;
 
   /**
-   * Change the rate of the flow of time for the clock
+   * \sa sleep_until
```
Comment (MichaelOrlov) [inline rosbag2_cpp/include/rosbag2_cpp/clocks/player_clock.hpp:64]: What does it means tag `\sa`?

File: rosbag2_cpp/include/rosbag2_cpp/clocks/time_controller_clock.hpp [hunk 7]
```diff
@@ -70,11 +67,19 @@ class TimeControllerClock : public PlayerClock
...
   bool sleep_until(rcutils_time_point_value_t until) override;
 
+  ROSBAG2_CPP_PUBLIC
+  bool sleep_until(rclcpp::Time until) override;
+
   /**
-   * Change the rate of the flow of time for the clock
+   * Change the rate of the flow of time for the clock.
+   *
+   * \param rate tTe new rate of time
```
Comment (MichaelOrlov) [inline rosbag2_cpp/include/rosbag2_cpp/clocks/time_controller_clock.hpp:76]: Typo? `tTe`

File: rosbag2_cpp/test/rosbag2_cpp/test_time_controller_clock.cpp [hunk 8]
```diff
@@ -228,3 +236,20 @@ TEST_F(TimeControllerClockTest, change_rate)
...
+  clock.jump(expected_ros_time);
+  EXPECT_EQ(clock.now(), expected_ros_time);
+
+  return_time += std::chrono::seconds(10);
+  expected_ros_time += RCUTILS_S_TO_NS(10);
+  EXPECT_EQ(clock.now(), expected_ros_time);
+
+  clock.jump(rclcpp::Time(100, 0));
+  return_time += std::chrono::seconds(1);
+  EXPECT_EQ(clock.now(), RCUTILS_S_TO_NS(101));
```
Comment (MichaelOrlov) [inline rosbag2_cpp/test/rosbag2_cpp/test_time_controller_clock.cpp:254]: It would be nice to add one more case for jumping back in time.

File: rosbag2_cpp/test/rosbag2_cpp/test_time_controller_clock.cpp [hunk 9]
```diff
@@ -47,86 +47,94 @@ TEST_F(TimeControllerClockTest, must_provide_now_fn)
...
+  EXPECT_LT(std::abs(clock.now() - over_limit_nanos), 10LL);
 }
 
 TEST_F(TimeControllerClockTest, nonzero_start_time)
 {
   ros_start_time = 1234567890LL;
-  const double playback_rate = 1.0;
-  rosbag2_cpp::TimeControllerClock pclock(ros_start_time, playback_rate, now_fn);
+  rosbag2_cpp::TimeControllerClock clock(ros_start_time, now_fn);
+  clock.set_rate(1.0);
```
Comment (MichaelOrlov) [inline rosbag2_cpp/test/rosbag2_cpp/test_time_controller_clock.cpp:88]: I would add 
```cpp
EXPECT_TRUE(clock.set_rate(1.0)); 
```
And somewhere 
```cpp
EXPECT_FALSE(clock.set_rate(0.0));
EXPECT_FALSE(clock.set_rate(-0.5));
```

**Review by MichaelOrlov — COMMENTED** [review]
@emersonknapp New version looks good to me among a few minor comments in tests and doxygen comments.

**Review by MichaelOrlov — APPROVED** [review]

File: rosbag2_cpp/include/rosbag2_cpp/clocks/player_clock.hpp (hunk 6 above)
Comment (Karsten1987) [reply rosbag2_cpp/include/rosbag2_cpp/clocks/player_clock.hpp:64]: I believe it stands for `see also`. https://www.doxygen.nl/manual/commands.html#cmdsa

File: rosbag2_cpp/include/rosbag2_cpp/clocks/time_controller_clock.hpp [hunk 10]
```diff
@@ -70,11 +67,19 @@ class TimeControllerClock : public PlayerClock
...
+   * \return false if rate <= 0, true otherwise.
+   *   It is not currently valid behavior to move time backwards.
+   *   To stop time, \sa pause.
```
Comment (Karsten1987) [inline rosbag2_cpp/include/rosbag2_cpp/clocks/time_controller_clock.hpp:79]: nitpick, I'd put this before the `\return` as a general comment

File: rosbag2_cpp/test/rosbag2_cpp/test_time_controller_clock.cpp (hunk 9 above)
Comment (Karsten1987) [reply rosbag2_cpp/test/rosbag2_cpp/test_time_controller_clock.cpp:88]: I think that's done further down below, within "invalid rate"

File: rosbag2_transport/src/rosbag2_transport/player.cpp [hunk 11]
```diff
@@ -83,16 +83,6 @@ Player::Player(const std::string & node_name, const rclcpp::NodeOptions & node_o
...
-Player::Player(
-  const rosbag2_storage::StorageOptions & storage_options,
-  const rosbag2_transport::PlayOptions & play_options,
-  const std::string & node_name,
-  const rclcpp::NodeOptions & node_options)
-: Player(std::make_unique<rosbag2_cpp::Reader>(),
-    storage_options, play_options,
-    node_name, node_options)
-{}
-
```
Comment (Karsten1987) [inline rosbag2_transport/src/rosbag2_transport/player.cpp:95]: Why this change? This is currently being used within the rosbag2_bag_v2 plugins

**Review by Karsten1987 — CHANGES_REQUESTED** [review]
This looks good to me. A few nitpicks inline and I also agree with some comments of @MichaelOrlov to extend a few test cases. 

I do request changes though because I don't understand why you deleted a constructor in the `Player`. Did I miss something here?

File: rosbag2_transport/src/rosbag2_transport/player.cpp (hunk 11 above)
Comment (emersonknapp) [reply rosbag2_transport/src/rosbag2_transport/player.cpp:95]: Oh - this was a bad rebase - good catch

File: rosbag2_cpp/test/rosbag2_cpp/test_time_controller_clock.cpp (hunk 9 above)
Comment (emersonknapp) [reply rosbag2_cpp/test/rosbag2_cpp/test_time_controller_clock.cpp:88]: Yep, it's a separate test

**Comment by emersonknapp** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14500/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9259/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12167/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14601/)

**Review by Karsten1987 — APPROVED** [review]

File: rosbag2_cpp/test/rosbag2_cpp/test_time_controller_clock.cpp (hunk 9 above)
Comment (MichaelOrlov) [reply rosbag2_cpp/test/rosbag2_cpp/test_time_controller_clock.cpp:88]: Sorry I overlooked it.

**Review by MichaelOrlov — APPROVED** [review]
:+1:
